File: macjet/ui/reclaim_panel.py
"""
MacJet — Reclaim Panel (Kill List)
Scored recommendation view showing which processes to kill and why.
"""

from __future__ import annotations

from textual.app import ComposeResult
from textual.widget import Widget
from textual.widgets import DataTable, Static

from ..collectors.metrics_history import ReclaimCandidate


# ─── Color ramps for score ───────────────────────────
def _score_color(score: int) -> str:
    if score >= 80:
        return "#FF4D6D"
    elif score >= 60:
        return "#FF8A4C"
    elif score >= 40:
        return "#FDBA35"
    elif score >= 20:
        return "#A78BFA"
    return "#7F8DB3"


def _risk_badge(risk: str) -> str:
    colors = {
        "safe": "[#32D583]SAFE[/]",
        "review": "[#FDBA35]REVIEW[/]",
        "danger": "[#FF4D6D]DANGER[/]",
    }
    return colors.get(risk, risk)


def _format_mem(mb: float) -> str:
    if mb >= 1024:
        return f"{mb / 1024:.1f}GB"
    return f"{mb:.0f}MB"


class ReclaimPanel(Widget):
    """Kill List panel showing scored process recommendations."""

    DEFAULT_CSS = """
    ReclaimPanel {
        height: 1fr;
        background: #0A0F1E;
    }
    ReclaimPanel #reclaim-toolbar {
        height: 1;
        background: #10182B;
        border-bottom: solid #1A2540;
        padding: 0 1;
        color: #7F8DB3;
    }
    ReclaimPanel DataTable {
        height: 1fr;
    }
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._table: DataTable | None = None
        self._candidates: list[ReclaimCandidate] = []
        self._row_keys: list[str] = []

    def compose(self) -> ComposeResult:
        yield Static(
            "  RECLAIM  [#7F8DB3]Scored recommendations • q:Quit App • k:Kill • K:Force Kill[/]",
            id="reclaim-toolbar",
        )
        table = DataTable(id="reclaim-table", cursor_type="row")
        table.add_column("", key="rail", width=1)
        table.add_column("Score", key="score", width=6)
        table.add_column("App", key="app", width=24)
        table.add_column("Reclaim", key="reclaim", width=16)
        table.add_column("Risk", key="risk", width=8)
        table.add_column("Reason", key="reason", width=36)
        table.add_column("Action", key="action", width=14)
        self._table = table
        yield table

    def update_data(self, candidates: list[ReclaimCandidate]):
        """Update the reclaim table with scored candidates."""
        if not self._table:
            return

        self._candidates = candidates
        table = self._table

        # Save cursor
        saved_row = table.cursor_row

        table.clear()
        self._row_keys = []

        for candidate in candidates:
            if candidate.score < 5:
                continue  # Skip very low scores

            # Severity rail
            sc = _score_color(candidate.score)
            rail = f"[{sc}]█[/]"

            # Score
            score_str = f"[{sc}]{candidate.score:3d}[/]"

            # App name
            app_str = f"{candidate.icon} {candidate.app_name}"
            if len(app_str) > 23:
                app_str = app_str[:22] + "…"

            # Reclaim estimate
            cpu_str = f"{candidate.reclaim_cpu:.0f}%"
            mem_str = _format_mem(candidate.reclaim_mem_mb)
            reclaim_str = f"~{cpu_str} / {mem_str}"

            # Risk badge
            risk_str = _risk_badge(candidate.risk)

            # Reason (truncate)
            reason = candidate.reason
            if len(reason) > 35:
                reason = reason[:34] + "…"

            # Action
            action_str = candidate.suggested_action

            row_key = f"reclaim-{candidate.group_key}"
            table.add_row(
                rail,
                score_str,
                app_str,
                reclaim_str,
                risk_str,
                reason,
                action_str,
                key=row_key,
            )
            self._row_keys.append(row_key)

        # Restore cursor
        if saved_row < len(self._row_keys):
            table.move_cursor(row=saved_row)

    def get_selected_group_key(self) -> str | None:
        """Get the group key of the currently selected reclaim candidate."""
        if not self._table or not self._row_keys:
            return None
        cursor_row = self._table.cursor_row
        if 0 <= cursor_row < len(self._row_keys):
            key = self._row_keys[cursor_row]
            if key.startswith("reclaim-"):
                return key[8:]
        return None

    def get_selected_candidate(self) -> ReclaimCandidate | None:
        """Get the currently selected ReclaimCandidate."""
        if not self._table or not self._candidates:
            return None
        cursor_row = self._table.cursor_row
        # Candidates and rows should be 1:1 (excluding <5 scores)
        visible = [c for c in self._candidates if c.score >= 5]
        if 0 <= cursor_row < len(visible):
            return visible[cursor_row]
        return None

File: src/ui/reclaim_panel.rs
/// MacJet — Reclaim Panel (Kill List)
/// Scored recommendation view showing which processes to kill and why.
use crate::collectors::metrics_history::ReclaimCandidate;
use ratatui::{
    layout::{Constraint, Rect},
    style::{Color, Modifier, Style},
    text::Span,
    widgets::{Block, Borders, Cell, Row, Table, TableState},
};

pub struct ReclaimPanelWidget<'a> {
    pub candidates: &'a [ReclaimCandidate],
    pub table_state: &'a mut TableState,
}

impl<'a> ReclaimPanelWidget<'a> {
    pub fn new(candidates: &'a [ReclaimCandidate], table_state: &'a mut TableState) -> Self {
        Self {
            candidates,
            table_state,
        }
    }

    fn score_color(score: u8) -> Color {
        if score >= 80 {
            Color::Rgb(255, 77, 109) // Danger
        } else if score >= 60 {
            Color::Rgb(255, 138, 76) // Warning
        } else if score >= 40 {
            Color::Rgb(253, 186, 53) // Orange
        } else if score >= 20 {
            Color::Rgb(167, 139, 250) // Purple
        } else {
            Color::Rgb(127, 141, 179) // Dim
        }
    }

    fn risk_badge(risk: &str) -> Span<'static> {
        match risk.to_lowercase().as_str() {
            "safe" => Span::styled("SAFE", Style::default().fg(Color::Rgb(50, 213, 131))),
            "review" => Span::styled("REVIEW", Style::default().fg(Color::Rgb(253, 186, 53))),
            "danger" => Span::styled("DANGER", Style::default().fg(Color::Rgb(255, 77, 109))),
            _ => Span::raw(risk.to_string()),
        }
    }

    fn format_mem(mb: f64) -> String {
        if mb >= 1024.0 {
            format!("{:.1}GB", mb / 1024.0)
        } else {
            format!("{:.0}MB", mb)
        }
    }
}

impl<'a> ratatui::widgets::Widget for ReclaimPanelWidget<'a> {
    fn render(self, area: Rect, buf: &mut ratatui::buffer::Buffer) {
        let block = Block::default()
            .borders(Borders::NONE)
            .style(Style::default().bg(Color::Rgb(10, 15, 30)));

        let header_cells = vec![
            Cell::from(""),
            Cell::from("Score"),
            Cell::from("App"),
            Cell::from("Reclaim"),
            Cell::from("Risk"),
            Cell::from("Reason"),
            Cell::from("Action"),
        ];

        let header = Row::new(header_cells)
            .style(
                Style::default()
                    .fg(Color::Rgb(127, 141, 179))
                    .bg(Color::Rgb(16, 24, 43)),
            )
            .height(1)
            .bottom_margin(0);

        let mut rows = Vec::new();
        for (idx, candidate) in self.candidates.iter().filter(|c| c.score >= 5).enumerate() {
            let sc = Self::score_color(candidate.score);
            let rail = Span::styled("█", Style::default().fg(sc));
            let score_str = Span::styled(format!("{:3}", candidate.score), Style::default().fg(sc));

            let mut app_str = format!("{} {}", candidate.icon, candidate.app_name);
            if app_str.chars().count() > 23 {
                app_str = format!("{}…", app_str.chars().take(22).collect::<String>());
            }

            let reclaim_str = format!(
                "~{:.0}% / {}",
                candidate.reclaim_cpu,
                Self::format_mem(candidate.reclaim_mem_mb)
            );
            let risk_str = Self::risk_badge(candidate.risk.as_str());

            let mut reason = candidate.reason.to_string();
            if reason.chars().count() > 35 {
                reason = format!("{}…", reason.chars().take(34).collect::<String>());
            }

            let cells = vec![
                Cell::from(rail),
                Cell::from(score_str),
                Cell::from(app_str).style(Style::default().fg(Color::Rgb(230, 236, 255))),
                Cell::from(reclaim_str).style(Style::default().fg(Color::Rgb(230, 236, 255))),
                Cell::from(risk_str),
                Cell::from(reason).style(Style::default().fg(Color::Rgb(230, 236, 255))),
                Cell::from(candidate.suggested_action.as_str())
                    .style(Style::default().fg(Color::Rgb(96, 165, 250))),
            ];

            let base_bg = if idx % 2 == 1 {
                Color::Rgb(14, 20, 37) // #0E1425
            } else {
                Color::Rgb(10, 15, 30) // #0A0F1E
            };

            rows.push(
                Row::new(cells)
                    .style(Style::default().bg(base_bg))
                    .height(1),
            );
        }

        let table = Table::new(
            rows,
            &[
                Constraint::Length(1),
                Constraint::Length(6),
                Constraint::Length(24),
                Constraint::Length(16),
                Constraint::Length(8),
                Constraint::Length(36),
                Constraint::Length(14),
            ],
        )
        .header(header)
        .block(block)
        .row_highlight_style(
            Style::default()
                .fg(Color::Rgb(96, 165, 250))
                .bg(Color::Rgb(18, 59, 87))
                .add_modifier(Modifier::BOLD),
        ) // #123B57 cursor
        .highlight_symbol("");

        // Use standard StatefulWidget rendering pattern directly
        ratatui::widgets::StatefulWidget::render(table, area, buf, self.table_state);
    }
}
